File: src/main/java/fi/metatavu/metaform/server/rest/translate/MetaformTranslator.java
package fi.metatavu.metaform.server.rest.translate;

import java.io.IOException;

import javax.enterprise.context.ApplicationScoped;
import javax.inject.Inject;

import fi.metatavu.metaform.server.persistence.model.Metaform;
import org.slf4j.Logger;

import com.fasterxml.jackson.databind.ObjectMapper;


/**
 * Translatro for Metaforms
 * 
 * @author Antti Leppä
 */
@ApplicationScoped
public class MetaformTranslator {
  
  @Inject
  private Logger logger;

  /**
   * Translates JPA Metaform into REST Metaform
   * 
   * @param entity JPA Metaform
   * @return REST Metaform
   */
  public fi.metatavu.metaform.api.spec.model.Metaform translateMetaform(fi.metatavu.metaform.server.persistence.model.Metaform entity) {
    if (entity == null) {
      return null;
    }

    fi.metatavu.metaform.api.spec.model.Metaform result = null;
    
    ObjectMapper objectMapper = new ObjectMapper();
    
    try {
      result = objectMapper.readValue(entity.getData(), fi.metatavu.metaform.api.spec.model.Metaform.class);
    } catch (IOException e) {
      logger.error(String.format("Failed to translate metaform %s", entity.getId().toString()), e);
      return null;
    }
    
    if (result != null) {
      result.setId(entity.getId());
      
      if (entity.getExportTheme() != null) {
        result.setExportThemeId(entity.getExportTheme().getId());
      }
    }
    
    return result;
  }
  
}

File: src/main/kotlin/fi/metatavu/metaform/server/rest/translate/MetaformTranslator.kt
package fi.metatavu.metaform.server.rest.translate

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import fi.metatavu.metaform.server.exceptions.MalformedMetaformJsonException
import fi.metatavu.metaform.server.persistence.dao.MetaformScriptDAO
import fi.metatavu.metaform.server.persistence.model.Metaform
import java.io.IOException
import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject

/**
 * Translator for Metaforms
 *
 * @author Antti Leppä
 */
@ApplicationScoped
class MetaformTranslator {

  @Inject
  lateinit var metaformScriptDAO: MetaformScriptDAO

  /**
   * Translates JPA Metaform into REST Metaform
   *
   * @param entity JPA Metaform
   * @return REST Metaform
   */
  @Throws(MalformedMetaformJsonException::class)
  fun translate(entity: Metaform): fi.metatavu.metaform.api.spec.model.Metaform {
    val objectMapper = ObjectMapper()
    objectMapper.registerModule(JavaTimeModule())

    val result = try {
       objectMapper.readValue(entity.data, fi.metatavu.metaform.api.spec.model.Metaform::class.java)
    } catch (e: IOException) {
      throw MalformedMetaformJsonException(String.format("Failed to translate metaform %s", entity.id.toString()), e)
    }

    val scripts = metaformScriptDAO.listByMetaform(entity).mapNotNull { metaformScript -> metaformScript.script.id}

    return result.copy(
      id = entity.id,
      slug = entity.slug,
      scripts = scripts,
      exportThemeId = entity.exportTheme?.id,
      visibility = entity.visibility,
      createdAt = entity.createdAt,
      modifiedAt = entity.modifiedAt,
      creatorId = entity.creatorId,
      lastModifierId = entity.lastModifierId
    )
  }
}